### PR TITLE
Add incentive for neo holders

### PR DESCRIPTION
**Pull request by Tommo-L**
The additional GAS allocation is as follows:

- **10% will be distributed to all NEO holders according to the weight of NEO. (Implemented by this PR)**
- _5% divided equally among 21 members._
- _85% will be allocated to the voters who voted for the top 21 nodes. The voters who voted for a consensus node are assigned 2/28 of this part according to the voting weight, and the voters who voted for a committee member (non-consensus node) are assigned 1/28 of this part according to the voting weight._

### CODE REVIEW

File: src/neo/SmartContract/Native/Tokens/NeoToken.cs [hunk 1]
```diff
@@ -50,45 +54,53 @@ protected override void OnBalanceChanging(ApplicationEngine engine, UInt160 acco
...
+
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_GasPerBlock), new StorageItem
+            {
+                Value = (5 * GAS.Factor).ToByteArray()
+            });
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_RewardRatio), new StorageItem(new RewardRatio
+            {
+                NeoHolder = 10,
+                Committee = 5,
+                Voter = 85
```
Comment (erikzhang) [inline src/neo/SmartContract/Native/Tokens/NeoToken.cs:101]: These values should be constants and cannot be modified.

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs:101]: Can't committee modify them?

Comment (erikzhang) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs:101]: No. They must be hard coded.

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs]: Done.

File: src/neo/SmartContract/Native/Tokens/NeoToken.cs [hunk 2]
```diff
@@ -50,45 +55,28 @@ protected override void OnBalanceChanging(ApplicationEngine engine, UInt160 acco
...
-                {
-                    amount += (Blockchain.DecrementInterval - istart) * Blockchain.GenerationAmount[ustart];
-                    ustart++;
-                    istart = 0;
-                }
-                amount += (iend - istart) * Blockchain.GenerationAmount[ustart];
-            }
-            return value * amount * GAS.Factor / TotalAmount;
+            var gasPerBlock = GetGasPerBlock(snapshot);
+            return value * gasPerBlock * (end - start) * NeoHolderRewardRatio / 100 / TotalAmount;
```
Comment (erikzhang) [inline src/neo/SmartContract/Native/Tokens/NeoToken.cs:68]: The value of `GasPerBlock` is read from storage. What if the value changed?

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs:68]: If we want to allow `GetGasPerBlock` to be modified, we need to record the historical value of `GasPerBlock`, or the incremental value. If so, we'd better add `SetGasPerBlock` method.

Comment (erikzhang) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs:68]: The committee has the permission to invoke `SetGasPerBlock`. And there is a range restriction on the value from `0` to `10` GAS pre block.

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs:68]: Ok, I'll add it.

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs]: Done.

File: src/neo/SmartContract/Native/Tokens/NeoToken.cs [hunk 3]
```diff
@@ -50,45 +55,48 @@ protected override void OnBalanceChanging(ApplicationEngine engine, UInt160 acco
...
         {
             if (value.IsZero || start >= end) return BigInteger.Zero;
             if (value.Sign < 0) throw new ArgumentOutOfRangeException(nameof(value));
-            BigInteger amount = BigInteger.Zero;
-            uint ustart = start / Blockchain.DecrementInterval;
-            if (ustart < Blockchain.GenerationAmount.Length)
+
+            StorageKey keyLeft = CreateStorageKey(Prefix_GasPerBlock).Add(uint.MaxValue - end);
+            StorageKey keyRight = CreateStorageKey(Prefix_GasPerBlock).Add(uint.MaxValue);
+            var enumerator = snapshot.Storages.FindRange(keyLeft, keyRight).GetEnumerator();
```
Comment (erikzhang) [inline src/neo/SmartContract/Native/Tokens/NeoToken.cs:70]: The integer in the key is little-endian. So you can't compare the keys. Maybe you can add a `AddBigEndian()` to replace `Add()`.

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs]: Good catch, fixed.

File: src/neo/SmartContract/Native/Tokens/NeoToken.cs [hunk 4]
```diff
@@ -100,13 +108,34 @@ protected override void OnPersist(ApplicationEngine engine)
             storage.Value = GetValidators(engine.Snapshot).ToByteArray();
         }
 
+        [ContractMethod(0_05000000, CallFlags.AllowModifyStates)]
+        private bool SetGasPerBlock(ApplicationEngine engine, BigInteger gasPerBlock)
+        {
+            if (gasPerBlock < 0 || gasPerBlock > 10 * GAS.Factor) return false;
+            if (!CheckCommittee(engine)) return false;
+            StorageKey key = CreateStorageKey(Prefix_GasPerBlock).AddBigEndian(uint.MaxValue - engine.Snapshot.PersistingBlock.Index - 1);
```
Comment (erikzhang) [inline src/neo/SmartContract/Native/Tokens/NeoToken.cs:116]: Maybe the key should be the index of the block. I know the performance of `next()` is higher than `prev()`, but it is not worth such optimization here.

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs:116]: No, and it's not very convenient to use the `DataCache.Seek` method.

If we want to seek(index = 10) in db, which contains 0, 8, 15.  It'll set position at 15.

Comment (erikzhang) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs:116]: What about we save all values in a single entry? I think it won't be modified often.

Comment (tornshares) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs:116]: It's ok.

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs:116]: It's good, I'll change it.

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs]: Adjusted, but It doesn't look as good as before.

**Review by erikzhang — APPROVED** [review]

**Review by erikzhang — APPROVED** [review]

**Comment by Tommo-L** [comment]
> CommitteeRewardRatio and VoterRewardRatio Bonus will come next?

Yes, there are in the next two PRs.

**Comment by cloud8little** [comment]
@erikzhang  @Tommo-L  will the genesis block(index=0) generate gas? if yes, will it also be used to reward the committees/neo holders?

**Pull request merged**